rust/yarp-sys/build.rs

```diff
@@ -0,0 +1,180 @@
+}
+
+fn cargo_manifest_path() -> &'static Path {
+    Path::new(env!("CARGO_MANIFEST_DIR"))
+}
+
+/// Uses `bindgen` to generate bindings to the C API. Update this to allow new types/functions/etc
+/// to be generated (it's allowlisted to only expose functions that'd make sense for public
+/// consumption).
+///
+/// This method only generates code in memory here--it doesn't write it to file.
+///
+fn generate_bindings(ruby_include_path: &Path) -> bindgen::Bindings {
+    bindgen::Builder::default()
+        .header(ruby_include_path.join("yarp/defines.h").to_string_lossy())
+        .header(ruby_include_path.join("yarp.h").to_string_lossy())
+        .clang_arg(format!("-I{}", ruby_include_path.display()))
+        .parse_callbacks(Box::new(bindgen::CargoCallbacks))
+        // Structs
+        .allowlist_type("yp_buffer_t") // TEST ✅
+        .allowlist_type("yp_comment_t") // TEST ✅
+        .allowlist_type("yp_diagnostic_t") // TEST ✅
+        .allowlist_type("yp_encoding_changed_callback_t") // TEST ✅
+        .allowlist_type("yp_encoding_decode_callback_t") // TEST ✅
+        .allowlist_type("yp_memsize_t") // TEST ✅
+        .allowlist_type("yp_list_t") // TEST ✅
+        .allowlist_type("yp_node_t") // TEST ✅
+        .allowlist_type("yp_parser_t") // TEST ✅
+        .allowlist_type("yp_pack_size") // TEST ✅
+        // TODO: Commenting this because I can't figure out how to get bindgen to generate the
+        // inner-unions nicely. Hand-rolling this in src/lib.rs for now.
+        // .allowlist_type("yp_string_t")
+        // TODO bindgen is still generating the inner unions, despite this...
+        .blocklist_type("^yp_string_t*") // TEST ✅
+        .allowlist_type("yp_string_list_t") // TEST ✅
+        .allowlist_type("yp_token_type_t")
+        // Enums
+        .rustified_non_exhaustive_enum("yp_comment_type_t") // TEST ✅
+        .rustified_non_exhaustive_enum("yp_context_t")
+        .rustified_non_exhaustive_enum("yp_heredoc_indent_t")
+        .rustified_non_exhaustive_enum("yp_heredoc_quote_t")
+        .rustified_non_exhaustive_enum("yp_lex_mode_t")
+        .rustified_non_exhaustive_enum("yp_lex_state_t")
+        .rustified_non_exhaustive_enum("yp_node_type_t") // TEST ✅
+        .rustified_non_exhaustive_enum("yp_pack_encoding") // TEST ✅
+        .rustified_non_exhaustive_enum("yp_pack_endian") // TEST ✅
+        .rustified_non_exhaustive_enum("yp_pack_length_type") // TEST ✅
+        .rustified_non_exhaustive_enum("yp_pack_result") // TEST ✅
+        .rustified_non_exhaustive_enum("yp_pack_signed") // TEST ✅
+        .rustified_non_exhaustive_enum("yp_pack_size") // TEST ✅
+        .rustified_non_exhaustive_enum("yp_pack_type") // TEST ✅
+        .rustified_non_exhaustive_enum("yp_pack_variant") // TEST ✅
+        .rustified_non_exhaustive_enum("yp_token_type")
+        .rustified_non_exhaustive_enum("yp_unescape_type_t") // TEST ✅
+        // Functions
+        .allowlist_function("yp_buffer_init") // TEST ✅
```

Comment:
Again, the whitelisted/allowlisted functions here should be all the `YP_EXPORTED_FUNCTION`s; that was my intent, at least. I'm happy to trim it down or add to it if that'd be better.